course02/project01/gavin/Asteroids/laser.cc

```diff
@@ -0,0 +1,25 @@
+//
+// Created by gavinparker on 10-2-19.
+//
+
+#include <cinder/gl/gl.h>
+#include "laser.h"
+
+Laser::Laser(GameObject& parent, glm::vec2 direction, glm::vec2 position) : Collidable(parent, Tag::Laser, position, 1), mDirection(direction){}
+
+void Laser::Update(const float frameDelta)
+{
+    auto dist = mDirection * speed * frameDelta;
+    mPosition += dist;
+}
+
+void Laser::Draw()
+{
+    ci::gl::drawSolidCircle(mPosition, mSize);
+}
+
+void Laser::Collide(Collidable &other) {
+    if(other.GetTag() == Tag::Asteroid)
```

Comment:
What I really want to express here is a method that takes any 'Collidable' but that can be specialised for a subset of types derived from Collidable. I don't know if that's possible without templating every Collidable with a derived type and adding it as a field, which is all I could think of.

Reply:
See my general comment: what you want here is just overloading. You want a set of `Collide(Asteroid,Laser)`, `Collide(Laser,Ship)`, etc that define the behavior for every possible collision. This way you preserve the real type, which is good because the collision code is _specific_ to the types - you thus don't want polymorphism/genericity here.

Reply:
These are now overloads in 76db252356e796ace0396de1cb5a68c9790e3d06, however i've kept the ```Collidable``` class to keep some logic out of ```GameObject```